### PR TITLE
[BugFix] check if helper in local before bdb setup

### LINKED ISSUE #7498: FE master crash unexpectedly
### Steps to reproduce the behavior (Required)
1. Setup a three-node FE cluster
2. Add an FE node but without `helper` mode

### Expected behavior (Required)
Add the FE node failed.

### Real behavior (Required)
But the FE master crashed with following log:

```java
2022-06-20 16:29:32,757 ERROR (heartbeat mgr|33) [BDBEnvironment.getDatabaseNames():351] bdb environment failure exception.
com.sleepycat.je.EnvironmentFailureException: (JE 7.3.7) Environment must be closed, caused by: com.sleepycat.je.EnvironmentFailureException: Environment invalid because of previous exception: (JE 7.3.7) 10.10.94.95_9010_1637139054525(3):/opt/starrocks/fe/doris-meta/bdb com.sleepycat.je.EnvironmentFailureException: (JE 7.3.7) Ids: 1 were equal. But names: 10.10.94.93_9010_1637138716667, 10.10.94.94_9010_1655704310986 weren't! UNEXPECTED_STATE: Unexpected internal state, may have side effects. Uncaught exception in feeder thread:Thread[Feeder Input,5,main] Originally thrown by HA thread: MASTER 10.10.94.95_9010_1637139054525(3) UNCAUGHT_EXCEPTION: Uncaught Exception in internal thread, unable to continue. Environment is invalid and must be closed.
	at com.sleepycat.je.EnvironmentFailureException.wrapSelf(EnvironmentFailureException.java:228) ~[je-7.3.7.jar:7.3.7]
	at com.sleepycat.je.dbi.EnvironmentImpl.checkIfInvalid(EnvironmentImpl.java:1766) ~[je-7.3.7.jar:7.3.7]
	at com.sleepycat.je.dbi.EnvironmentImpl.checkOpen(EnvironmentImpl.java:1775) ~[je-7.3.7.jar:7.3.7]
	at com.sleepycat.je.Environment.checkOpen(Environment.java:2473) ~[je-7.3.7.jar:7.3.7]
	at com.sleepycat.je.Environment.getDatabaseNames(Environment.java:2245) ~[je-7.3.7.jar:7.3.7]
	at com.starrocks.journal.bdbje.BDBEnvironment.getDatabaseNames(BDBEnvironment.java:344) [starrocks-fe.jar:?]
	at com.starrocks.journal.bdbje.BDBJEJournal.getMaxJournalId(BDBJEJournal.java:209) [starrocks-fe.jar:?]
	at com.starrocks.persist.EditLog.getMaxJournalId(EditLog.java:98) [starrocks-fe.jar:?]
	at com.starrocks.system.HeartbeatMgr.runAfterCatalogReady(HeartbeatMgr.java:158) [starrocks-fe.jar:?]
	at com.starrocks.common.util.MasterDaemon.runOneCycle(MasterDaemon.java:61) [starrocks-fe.jar:?]
	at com.starrocks.common.util.Daemon.run(Daemon.java:119) [starrocks-fe.jar:?]
Caused by: com.sleepycat.je.EnvironmentFailureException: Environment invalid because of previous exception: (JE 7.3.7) 10.10.94.95_9010_1637139054525(3):/opt/starrocks/fe/doris-meta/bdb com.sleepycat.je.EnvironmentFailureException: (JE 7.3.7) Ids: 1 were equal. But names: 10.10.94.93_9010_1637138716667, 10.10.94.94_9010_1655704310986 weren't! UNEXPECTED_STATE: Unexpected internal state, may have side effects. Uncaught exception in feeder thread:Thread[Feeder Input,5,main] Originally thrown by HA thread: MASTER 10.10.94.95_9010_1637139054525(3) UNCAUGHT_EXCEPTION: Uncaught Exception in internal thread, unable to continue. Environment is invalid and must be closed.
	at com.sleepycat.je.utilint.StoppableThread.handleUncaughtException(StoppableThread.java:205) ~[je-7.3.7.jar:7.3.7]
	at com.sleepycat.je.utilint.StoppableThread$UncaughtHandler.uncaughtException(StoppableThread.java:236) ~[je-7.3.7.jar:7.3.7]
	at java.lang.Thread.dispatchUncaughtException(Thread.java:1959) ~[?:1.8.0_221]
Caused by: com.sleepycat.je.EnvironmentFailureException: (JE 7.3.7) Ids: 1 were equal. But names: 10.10.94.93_9010_1637138716667, 10.10.94.94_9010_1655704310986 weren't! UNEXPECTED_STATE: Unexpected internal state, may have side effects. Uncaught exception in feeder thread:Thread[Feeder Input,5,main] Originally thrown by HA thread: MASTER 10.10.94.95_9010_1637139054525(3)
	at com.sleepycat.je.EnvironmentFailureException.unexpectedState(EnvironmentFailureException.java:428) ~[je-7.3.7.jar:7.3.7]
	at com.sleepycat.je.rep.impl.node.NameIdPair.equals(NameIdPair.java:165) ~[je-7.3.7.jar:7.3.7]
	at com.sleepycat.je.rep.impl.RepGroupImpl.checkForConflicts(RepGroupImpl.java:362) ~[je-7.3.7.jar:7.3.7]
	at com.sleepycat.je.rep.impl.RepGroupDB$2.phase1Body(RepGroupDB.java:756) ~[je-7.3.7.jar:7.3.7]
	at com.sleepycat.je.rep.impl.RepGroupDB$TwoPhaseUpdate.phase1(RepGroupDB.java:1259) ~[je-7.3.7.jar:7.3.7]
	at com.sleepycat.je.rep.impl.RepGroupDB$TwoPhaseUpdate.execute(RepGroupDB.java:1342) ~[je-7.3.7.jar:7.3.7]
	at com.sleepycat.je.rep.impl.RepGroupDB.twoPhaseMemberUpdate(RepGroupDB.java:778) ~[je-7.3.7.jar:7.3.7]
	at com.sleepycat.je.rep.impl.RepGroupDB.createMember(RepGroupDB.java:667) ~[je-7.3.7.jar:7.3.7]
	at com.sleepycat.je.rep.impl.RepGroupDB.ensureMember(RepGroupDB.java:578) ~[je-7.3.7.jar:7.3.7]
	at com.sleepycat.je.rep.impl.RepGroupDB.ensureMember(RepGroupDB.java:483) ~[je-7.3.7.jar:7.3.7]
	at com.sleepycat.je.rep.stream.FeederReplicaHandshake.verifyMembershipInfo(FeederReplicaHandshake.java:385) ~[je-7.3.7.jar:7.3.7]
	at com.sleepycat.je.rep.stream.FeederReplicaHandshake.execute(FeederReplicaHandshake.java:283) ~[je-7.3.7.jar:7.3.7]
	at com.sleepycat.je.rep.impl.node.Feeder$InputThread.run(Feeder.java:611) ~[je-7.3.7.jar:7.3.7]
2022-06-20 16:29:32,758 INFO (Thread-43|98) [StarRocksFE.lambda$addShutdownHook$1():144] start to execute shutdown hook
2022-06-20 16:29:32,764 INFO (Thread-10913|1247796) [BDBEnvironment.close():389] start to close log databases
2022-06-20 16:29:32,764 WARN (Thread-10913|1247796) [StarRocksFE.lambda$null$0():161] close edit log failed
com.sleepycat.je.EnvironmentFailureException: (JE 7.3.7) Environment must be closed, caused by: com.sleepycat.je.EnvironmentFailureException: Environment invalid because of previous exception: (JE 7.3.7) 10.10.94.95_9010_1637139054525(3):/opt/starrocks/fe/doris-meta/bdb com.sleepycat.je.EnvironmentFailureException: (JE 7.3.7) Ids: 1 were equal. But names: 10.10.94.93_9010_1637138716667, 10.10.94.94_9010_1655704310986 weren't! UNEXPECTED_STATE: Unexpected internal state, may have side effects. Uncaught exception in feeder thread:Thread[Feeder Input,5,main] Originally thrown by HA thread: MASTER 10.10.94.95_9010_1637139054525(3) UNCAUGHT_EXCEPTION: Uncaught Exception in internal thread, unable to continue. Environment is invalid and must be closed.
	at com.sleepycat.je.EnvironmentFailureException.wrapSelf(EnvironmentFailureException.java:228) ~[je-7.3.7.jar:7.3.7]
	at com.sleepycat.je.dbi.EnvironmentImpl.checkIfInvalid(EnvironmentImpl.java:1766) ~[je-7.3.7.jar:7.3.7]
	at com.sleepycat.je.dbi.EnvironmentImpl.checkOpen(EnvironmentImpl.java:1775) ~[je-7.3.7.jar:7.3.7]
	at com.sleepycat.je.Environment.checkOpen(Environment.java:2473) ~[je-7.3.7.jar:7.3.7]
	at com.sleepycat.je.Database.checkEnv(Database.java:2287) ~[je-7.3.7.jar:7.3.7]
	at com.sleepycat.je.Database.getDatabaseName(Database.java:2045) ~[je-7.3.7.jar:7.3.7]
	at com.starrocks.journal.bdbje.BDBEnvironment.close(BDBEnvironment.java:394) ~[starrocks-fe.jar:?]
	at com.starrocks.StarRocksFE.lambda$null$0(StarRocksFE.java:157) ~[starrocks-fe.jar:?]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_221]
Caused by: com.sleepycat.je.EnvironmentFailureException: Environment invalid because of previous exception: (JE 7.3.7) 10.10.94.95_9010_1637139054525(3):/opt/starrocks/fe/doris-meta/bdb com.sleepycat.je.EnvironmentFailureException: (JE 7.3.7) Ids: 1 were equal. But names: 10.10.94.93_9010_1637138716667, 10.10.94.94_9010_1655704310986 weren't! UNEXPECTED_STATE: Unexpected internal state, may have side effects. Uncaught exception in feeder thread:Thread[Feeder Input,5,main] Originally thrown by HA thread: MASTER 10.10.94.95_9010_1637139054525(3) UNCAUGHT_EXCEPTION: Uncaught Exception in internal thread, unable to continue. Environment is invalid and must be closed.
	at com.sleepycat.je.utilint.StoppableThread.handleUncaughtException(StoppableThread.java:205) ~[je-7.3.7.jar:7.3.7]
	at com.sleepycat.je.utilint.StoppableThread$UncaughtHandler.uncaughtException(StoppableThread.java:236) ~[je-7.3.7.jar:7.3.7]
	at java.lang.Thread.dispatchUncaughtException(Thread.java:1959) ~[?:1.8.0_221]
Caused by: com.sleepycat.je.EnvironmentFailureException: (JE 7.3.7) Ids: 1 were equal. But names: 10.10.94.93_9010_1637138716667, 10.10.94.94_9010_1655704310986 weren't! UNEXPECTED_STATE: Unexpected internal state, may have side effects. Uncaught exception in feeder thread:Thread[Feeder Input,5,main] Originally thrown by HA thread: MASTER 10.10.94.95_9010_1637139054525(3)
	at com.sleepycat.je.EnvironmentFailureException.unexpectedState(EnvironmentFailureException.java:428) ~[je-7.3.7.jar:7.3.7]
	at com.sleepycat.je.rep.impl.node.NameIdPair.equals(NameIdPair.java:165) ~[je-7.3.7.jar:7.3.7]
	at com.sleepycat.je.rep.impl.RepGroupImpl.checkForConflicts(RepGroupImpl.java:362) ~[je-7.3.7.jar:7.3.7]
	at com.sleepycat.je.rep.impl.RepGroupDB$2.phase1Body(RepGroupDB.java:756) ~[je-7.3.7.jar:7.3.7]
	at com.sleepycat.je.rep.impl.RepGroupDB$TwoPhaseUpdate.phase1(RepGroupDB.java:1259) ~[je-7.3.7.jar:7.3.7]
	at com.sleepycat.je.rep.impl.RepGroupDB$TwoPhaseUpdate.execute(RepGroupDB.java:1342) ~[je-7.3.7.jar:7.3.7]
	at com.sleepycat.je.rep.impl.RepGroupDB.twoPhaseMemberUpdate(RepGroupDB.java:778) ~[je-7.3.7.jar:7.3.7]
	at com.sleepycat.je.rep.impl.RepGroupDB.createMember(RepGroupDB.java:667) ~[je-7.3.7.jar:7.3.7]
	at com.sleepycat.je.rep.impl.RepGroupDB.ensureMember(RepGroupDB.java:578) ~[je-7.3.7.jar:7.3.7]
	at com.sleepycat.je.rep.impl.RepGroupDB.ensureMember(RepGroupDB.java:483) ~[je-7.3.7.jar:7.3.7]
	at com.sleepycat.je.rep.stream.FeederReplicaHandshake.verifyMembershipInfo(FeederReplicaHandshake.java:385) ~[je-7.3.7.jar:7.3.7]
	at com.sleepycat.je.rep.stream.FeederReplicaHandshake.execute(FeederReplicaHandshake.java:283) ~[je-7.3.7.jar:7.3.7]
	at com.sleepycat.je.rep.impl.node.Feeder$InputThread.run(Feeder.java:611) ~[je-7.3.7.jar:7.3.7]
2022-06-20 16:29:32,765 INFO (Thread-43|98) [StarRocksFE.lambda$addShutdownHook$1():174] shutdown hook end

[END] 2022-6-20 17:05:36
```

### StarRocks version (Required)
- 2.0.1

### PR DESCRIPTION
**Pull request by padmejin**
## What type of PR is this：
- [x] bug
- [ ] feature
- [ ] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #7498 

## Problem Summary(Required) ：

How to reproduce this problem:
1. Deploy a cluster with 1 master + 2 followers.
2. Restart the current master, wait until the whole cluster is stable. Let's say the new master is node A.
3. Add a new node X to the cluster, but start without `--helper` argument. Node X will restart successfully and consider itself as a master of its own cluster.
4. Restart the node X with `--helper` argument. 

We expect this time node X will be down for failure to join the cluster, but it turns out that node A is down. Which means that add a bad follower to a cluster could leading to master down.

The main reason for this bug is a bad design in the BDBJE library. When a node X joins an existing cluster, there is no mechanism to check if the cluster information in the metadata of node X matches the old ones such as node A.

Node X will vote for itself as the new master right after the connection is built, yet only when the id of node X is identical to node A will an exception throws at the side of node X. In other cases, node A will crash because of the conflict of ids when receives this proposal. Since node X is the first instance of its cluster, the id of node X is 1. But node A is not the first instance of its cluster, as a result node X successfully join the cluster and node A die of conflict of ids when handling the proposal from node X.

To solve this problem, we have to add a mechanism to check if local cluster matches the helper node. I add an extra step to init a observer environment so that we can read local nodes and check if they contain the helper node.

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 37 / 46 (80.43%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/journal/bdbje/BDBEnvironment.java|37|46|80.43%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 38 / 49 (77.55%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/journal/bdbje/BDBEnvironment.java|38|49|77.55%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 38 / 49 (77.55%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/journal/bdbje/BDBEnvironment.java|38|49|77.55%|

File: fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBEnvironment.java [hunk 1]
```diff
@@ -305,6 +307,7 @@ protected void setupEnvironment() throws JournalException, InterruptedException
                     refreshLog((InsufficientLogException) e);
                 }
                 close();
+                closing = false;
```
Comment (gengjun-git) [inline fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBEnvironment.java:310]: the closing=false has been set in the close(), why set it again here?

Comment (padmejin) [reply fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBEnvironment.java]: my mistake. fix it.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 37 / 48 (77.08%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/journal/bdbje/BDBEnvironment.java|37|48|77.08%|

**Review by gengjun-git — APPROVED** [review]

File: fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBEnvironment.java [hunk 2]
```diff
@@ -163,11 +164,12 @@ protected BDBEnvironment(File envHome, String selfNodeName, String selfNodeHostP
     // The setup() method opens the environment and database
     protected void setup() throws JournalException, InterruptedException {
         this.closing = false;
-        initConfigs();
+        ensureHelperInLocal();
+        initConfigs(isElectable);
```
Comment (chaoyli) [inline fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBEnvironment.java:168]: why pass the `isElectable` parameter?
This `isElectable` is not changed, you can use the member variable.

**Review by chaoyli — APPROVED** [review]

**Pull request merged**